flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/AvroToBigQuerySchemaTransform.java

```diff
@@ -0,0 +1,456 @@
+            // ['datatype, 'null'] and ['null', datatype] are only valid cases.
+            @Nullable org.apache.avro.Schema actualSchema = null;
+            if (unionTypes.get(0).getType() != org.apache.avro.Schema.Type.NULL
+                    && unionTypes.get(1).getType() == org.apache.avro.Schema.Type.NULL) {
+                actualSchema = unionTypes.get(0);
+            } else if (unionTypes.get(0).getType() == org.apache.avro.Schema.Type.NULL
+                    && unionTypes.get(1).getType() != org.apache.avro.Schema.Type.NULL) {
+                actualSchema = unionTypes.get(1);
+            }
+
+            /* UNION of type ARRAY is not supported.
+            ARRAY is mapped to REPEATED type in Bigquery, which cannot be NULLABLE.
+            If we have the datatype is ["null", "ARRAY"] or ["ARRAY", "null],
+            UnsupportedOperationException is thrown. */
+            if (actualSchema != null
+                    && actualSchema.getType() == org.apache.avro.Schema.Type.ARRAY) {
+                throw new UnsupportedOperationException(
+                        "NULLABLE ARRAYS in UNION types are not supported");
+            }
+
+            if (actualSchema != null) {
+                return convertAvroFieldToBigQueryField(actualSchema, name, nestedLevel)
+                        .toBuilder()
+                        .setMode(Field.Mode.NULLABLE)
+                        .build();
+            }
+        }
+        throw new IllegalArgumentException(
+                String.format(
+                        "Unsupported Avro Field: %s of type UNION. Only supported types are \"['datatype'], ['null', 'datatype'] or ['datatype', 'null']\"",
+                        name));
+    }
+
+    /**
+     * Converts an Avro ARRAY field to a BigQuery REPEATED field.
+     *
+     * <p>The following restrictions imposed by BigQuery's schema definition:
+     *
+     * <ul>
+     *   <li>Arrays of arrays are not supported.
+     *   <li>Array cannot have a NULLABLE element.
+     * </ul>
+     *
+     * <p>If any of these restrictions are violated, an {@link UnsupportedOperationException} is
+     * thrown.
+     *
+     * @param avroSchema The Avro schema of the repeated field.
+     * @param name The name of the field.
+     * @param nestedLevel The nesting level of the field.
+     * @return The converted BigQuery {@link Field} with mode set to {@link Field.Mode#REPEATED}.
+     * @throws UnsupportedOperationException if the Avro schema violates any of the restrictions for
+     *     BigQuery repeated fields.
+     */
+    private static Field convertAvroRepeatedFieldToBigQueryField(
+            org.apache.avro.Schema avroSchema, String name, int nestedLevel) {
+        if (avroSchema.getType() == org.apache.avro.Schema.Type.ARRAY) {
+            throw new IllegalStateException(
+                    String.format(
+                            "BigQuery ARRAY cannot have recursive ARRAY fields. Found recursive Array field: %s",
```

Comment:
Error messages do not seem consistent. Lets follow a uniform pattern across all IllegalArgument and UnsupportedOperation exceptions

Reply:
For all Illegal Argument Exceptions, the error message initially states: ```Could not convert avro field: %s of expected output to BigQuery table schema.``` and then the associated issue.